### PR TITLE
Feature/tests for controls

### PR DESCRIPTION
**Pull request by B3zo0**
- [x] Unit tests for all the fieldcontrols and viewcontrols
   - [x] AllowedChildtypes
   - [x] AutoComplete
   - [x] Avatar
   - [x] Boolean
   - [x] CheckboxGroup
   - [x] ColorPicker
   - [x] DatePicker
   - [x] DateTimePicker
   - [x] DropDownList
   - [x] EmptyFieldControl
   - [x] FileUpload
   - [x] Name
   - [x] Number
   - [x] Password
   - [x] RadioButtonGroup
   - [x] ReferenceGrid
   - [x] RichTextEditor
   - [x] ShortText
   - [x] TagsInput
   - [x] Textarea
   - [x] TimePicker
   - [x] NewView
   - [x] EditView
   - [x] BrowseView

- [x] Solve accessibility issues that are mentioned in storybook docs

### CODE REVIEW

File: packages/sn-controls-react/src/fieldcontrols/Avatar/AvatarPicker.tsx [hunk 1]
```diff
@@ -39,17 +39,15 @@ const pickerItemOptions: ODataParams<Folder> = {
  * Represents an avatar picker component
  */
 export function AvatarPicker(props: AvatarPickerProps) {
-  const { items, selectedItem, setSelectedItem, path, navigateTo, reload, isLoading, error } = useListPicker<
-    GenericContent
-  >({
+  const { items, selectedItem, setSelectedItem, path, navigateTo, reload, isLoading, error } = useListPicker<User>({
```
Comment (herflis) [inline packages/sn-controls-react/src/fieldcontrols/Avatar/AvatarPicker.tsx]: Avatar picker is totally empty (no list, no upload button) on admin-ui
image

File: packages/sn-controls-react/src/fieldcontrols/Boolean.tsx [hunk 2]
```diff
@@ -6,78 +6,47 @@ import FormControl from '@material-ui/core/FormControl'
...
+        <FormControl required={props.settings.Compulsory} disabled={props.settings.ReadOnly}>
+          <FormControlLabel
+            name={props.settings.Name}
+            control={<Checkbox checked={value} onChange={handleChange} />}
+            label={props.settings.DisplayName}
+          />
+          <FormHelperText>{props.settings.Description}</FormHelperText>
+        </FormControl>
+      )
+    case 'browse':
```
Comment (herflis) [inline packages/sn-controls-react/src/fieldcontrols/Boolean.tsx]: Boolean in browse mode is bit messed up
image

File: packages/sn-controls-react/src/fieldcontrols/FileName.tsx [hunk 3]
```diff
@@ -23,8 +23,8 @@ export function FileName(props: ReactClientFieldSetting) {
   const [value, setValue] = useState(valueInitialState)
 
   const getExtension = () => {
-    if (props.content && props.content[props.settings.Name] && props.content[props.settings.Name].indexOf('.') > -1) {
-      return props.content[props.settings.Name].substr(props.content[props.settings.Name].lastIndexOf('.') + 1)
+    if (props.fieldValue && props.fieldValue.indexOf('.') > -1) {
```
Comment (herflis) [inline packages/sn-controls-react/src/fieldcontrols/FileName.tsx]: If you try to update the name itself the extension is added to it on every keyup

image

File: packages/sn-controls-react/src/fieldcontrols/Number.tsx [hunk 4]
```diff
@@ -3,111 +3,91 @@
...
+          <Typography variant="caption" gutterBottom={true}>
+            {props.settings.DisplayName}
+          </Typography>
+          <Typography variant="body1" gutterBottom={true}>
+            {isCurrencyFieldSetting(props.settings) ? (props.settings.Format ? props.settings.Format : '$') : null}
+            {props.fieldValue}
+            {props.settings.ShowAsPercentage ? '%' : null}
+          </Typography>
+        </div>
+      ) : null
```
Comment (herflis) [inline packages/sn-controls-react/src/fieldcontrols/Number.tsx]: Maybe an example in storybook with double value and double value as Step would be nice.

File: packages/sn-controls-react/test/CheckboxGroup.test.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      ).toBe(defaultSettings.Options[0].Text)
+      expect(
+        wrapper
+          .find(FormControlLabel)
+          .last()
+          .prop('label'),
+      ).toBe(defaultSettings.Options[2].Text)
+    })
+
+    it('should show fieldValue it is passed as a string', () => {
```
Comment (herflis) [inline packages/sn-controls-react/test/CheckboxGroup.test.tsx]:
```suggestion
    it('should show fieldValue when it is passed as a string', () => {
```

File: packages/sn-controls-react/test/DropDownList.test.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+      ).toBe(defaultSettings.Options[0].Text)
+      expect(
+        wrapper
+          .find(FormControlLabel)
+          .last()
+          .prop('label'),
+      ).toBe(defaultSettings.Options[2].Text)
+    })
+
+    it('should show fieldValue it is passed as a string', () => {
```
Comment (herflis) [inline packages/sn-controls-react/test/DropDownList.test.tsx]:
```suggestion
    it('should show fieldValue when it is passed as a string', () => {
```

File: packages/sn-controls-react/test/Name.test.tsx [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+      expect(wrapper.find(TextField).prop('helperText')).toBe(defaultSettings.Description)
+      expect(wrapper).toMatchSnapshot()
+    })
+    it('should call on change when input changes', () => {
+      const fieldOnChange = jest.fn()
+      const wrapper = shallow(<Name actionName="edit" fieldOnChange={fieldOnChange} settings={defaultSettings} />)
+      wrapper.find(TextField).simulate('change', { target: { value: 'Hello World' } })
+      expect(fieldOnChange).toBeCalled()
+    })
+    it('should show error message when invalid character typed', () => {
```
Comment (herflis) [inline packages/sn-controls-react/test/Name.test.tsx]:
```suggestion
    it('should show error message when invalid character is typed', () => {
```

File: packages/sn-controls-react/test/ReferenceGrid.test.tsx [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+      updatedWrapper
+        .find(Dialog)
+        .find(Button)
+        .first()
+        .simulate('click')
+      expect(fieldOnChange).toBeCalled()
+      expect(updatedWrapper.find(DefaultItemTemplate)).toHaveLength(4)
+    })
+
+    it('should remove all the items when selected a new item with allow multiple false', async () => {
```
Comment (herflis) [inline packages/sn-controls-react/test/ReferenceGrid.test.tsx]:
```suggestion
    it('should remove all the items when a new item is selected with allow multiple false', async () => {
```

File: packages/sn-controls-react/src/fieldcontrols/FileName.tsx (hunk 3 above)
Comment (B3zo0) [reply packages/sn-controls-react/src/fieldcontrols/FileName.tsx]: fixed

File: packages/sn-controls-react/src/fieldcontrols/Boolean.tsx (hunk 2 above)
Comment (B3zo0) [reply packages/sn-controls-react/src/fieldcontrols/Boolean.tsx]: fixed

File: packages/sn-controls-react/src/fieldcontrols/Avatar/AvatarPicker.tsx (hunk 1 above)
Comment (B3zo0) [reply packages/sn-controls-react/src/fieldcontrols/Avatar/AvatarPicker.tsx]: fixed. now it falls back to the avatar url path when no upload folder path or selection root is available.

File: packages/sn-controls-react/src/fieldcontrols/FileName.tsx [hunk 9]
```diff
@@ -23,8 +23,8 @@ export function FileName(props: ReactClientFieldSetting) {
...
+      return props.fieldValue.substr(props.fieldValue.lastIndexOf('.') + 1)
```
[the remark this answers is not in the document]
Comment (B3zo0) [reply packages/sn-controls-react/src/fieldcontrols/FileName.tsx]: I've fixed that. it was a mistake I made 😨

**Pull request merged**